lib/Conversion/CalyxToFSM/CalyxToFSM.cpp

```diff
@@ -268,6 +268,74 @@ LogicalResult CompileFSMVisitor::visit(StateOp currentState, EnableOp enableOp,
   return success();
 }
 
+// CompileInvoke is used to convert invoke operations to group operations and
+// enable operations.
+class CompileInvoke {
+public:
+  CompileInvoke(ComponentOp &component, OpBuilder &builder)
+      : component(component), builder(builder) {}
+  void compile();
+
+private:
+  void lowerInvokeOp(InvokeOp &invokeOp);
+  ComponentOp &component;
+  OpBuilder &builder;
+  // It is used to pass to the go port and assign a value to the go port.
+  hw::ConstantOp constantOp = nullptr;
+  size_t label = 0;
+};
+
+// Access all invokeOp.
+void CompileInvoke::compile() {
+  auto invokeOps = component.getControlOp().getInvokeOps();
+  for (auto op : invokeOps) {
+    lowerInvokeOp(op);
+  }
+}
+
+// Convert an invoke operation to a group operation and an enable operation.
+void CompileInvoke::lowerInvokeOp(InvokeOp &invokeOp) {
+  if (!constantOp) {
+    WalkResult result = component.walk([&](hw::ConstantOp op) {
```

Comment:
What are you trying to achieve here? Reduce the number of constants used? IMO, that kind of folding can be done in a future pass.

Reply:
The first idea was to create a ConstantOp directly, and I also thought that maybe a later pass would fold the ConstantOp, but I'm not so sure, so I've done it here for the ConstantOp.